### PR TITLE
Workaround for vshl/vshr on aarch64 GCC

### PR DESCRIPTION
**Pull request by vorj**
related: #1815,  #1880

`vshl` / `vshr` of ARM NEON requires immediate (compiletime constant) value as shift parameter.
However, the implementations of those intrinsics on GCC can receive runtime value.
Current faiss implementation depends on this, so some correct-behavioring compilers like Clang can't build faiss for aarch64.
This PR fix this issue; thus faiss applied this PR can be built with Clang for aarch64 machines like M1 Mac.

### CODE REVIEW

**Comment by vsolovyov** [comment]
Wow, you're fast :) With this fix it does compile for me, and tests pass 100%.
Clang 12.0.5, macOS 11.3, M1 Mac

**Comment by mdouze** [comment]
Cool if it works. Hopefully the optimizer will be able to remove the switch at compile time.

File: faiss/utils/simdlib_neon.h [hunk 1]
```diff
@@ -252,14 +262,108 @@ struct simd16uint16 {
...
+                return simd16uint16{detail::simdlib::unary_func(
+                        data, detail::simdlib::vshrq<13>)};
+            case 14:
+                return simd16uint16{detail::simdlib::unary_func(
+                        data, detail::simdlib::vshrq<14>)};
+            case 15:
+                return simd16uint16{detail::simdlib::unary_func(
+                        data, detail::simdlib::vshrq<15>)};
+            default:
+                return *this;
```
Comment (beauby) [inline faiss/utils/simdlib_neon.h:312]: This assumes `(shift <= 1 || shift > 15)` means shift == 0, right? If so, we should probably have:
```
case 0:
  return *this;
default:
  FAISS_THROW_FMT("Invalid shift %d", shift);
```

Comment (vorj) [reply faiss/utils/simdlib_neon.h]: I've fixed them.

**Pull request closed without merging**